### PR TITLE
chore: Move some utility methods to submodules of scalar_funcs

### PR DESCRIPTION
**Pull request by advancedxy**
## Rationale for this change
This is a follow-up as discussed in https://github.com/apache/datafusion-comet/pull/449#discussion_r1608635628

## What changes are included in this PR?
1. hex_encode and wrap_digest_result_as_hex_string goes to hex submodule
2. spark_murmur3_hash and spark_xxhash64 goes to hash_expressions submodule
3. update benchmark and modify test code

## How are these changes tested?
Existing tests with one slightly modification.

### CODE REVIEW

File: core/src/execution/datafusion/expressions/scalar_funcs/hex.rs [hunk 1]
```diff
@@ -246,14 +291,14 @@ mod test {
     fn test_dictionary_hex_binary() {
         let mut input_builder = BinaryDictionaryBuilder::<Int32Type>::new();
         input_builder.append_value("1");
-        input_builder.append_value("1");
```
Comment (comphead) [inline core/src/execution/datafusion/expressions/scalar_funcs/hex.rs:249]: 👍

File: core/src/execution/datafusion/expressions/scalar_funcs/hash_expressions.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        _ => {
+            internal_err!(
+                "The seed of function murmur3_hash must be an Int32 scalar value, but got: {:?}.",
+                seed
+            )
+        }
+    }
+}
+
+pub fn spark_xxhash64(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
```
Comment (comphead) [inline core/src/execution/datafusion/expressions/scalar_funcs/hash_expressions.rs:67]: Its not this PR problem but we need a description to pub methods

Comment (tshauck) [reply core/src/execution/datafusion/expressions/scalar_funcs/hash_expressions.rs:67]: Additionally, might consider limiting the scope. This function probably isn't needed outside of `scalar_funcs`.

File: core/src/execution/datafusion/expressions/scalar_funcs/hash_expressions.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+use datafusion_expr::ColumnarValue;
+use std::sync::Arc;
+
+pub fn spark_murmur3_hash(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
+    let length = args.len();
+    let seed = &args[length - 1];
+    match seed {
+        ColumnarValue::Scalar(ScalarValue::Int32(Some(seed))) => {
+            // iterate over the arguments to find out the length of the array
+            let num_rows = args[0..args.len() - 1]
```
Comment (comphead) [inline core/src/execution/datafusion/expressions/scalar_funcs/hash_expressions.rs:30]: any chance here to be an index out of bounds?

File: core/src/execution/datafusion/expressions/scalar_funcs/hex.rs [hunk 4]
```diff
@@ -26,24 +26,69 @@ use arrow_schema::DataType;
...
+    hex_encode(data, true)
+}
+
+#[inline(always)]
+fn hex_bytes<T: AsRef<[u8]>>(bytes: T) -> Result<String, std::fmt::Error> {
+    let hex_string = hex_encode(bytes, false);
     Ok(hex_string)
 }
 
+pub(super) fn wrap_digest_result_as_hex_string(
```
Comment (tshauck) [inline core/src/execution/datafusion/expressions/scalar_funcs/hex.rs:66]: A bit of personal opinion, but while this is doing hexing, IMO because it's related to the hash scalar function, it'd be more appropriate to put it there.

**Review by tshauck — APPROVED** [review]
Other's probably have more substantive feedback, but in terms of a good step for better organization, this LGTM

File: core/benches/hash.rs [hunk 5]
```diff
@@ -100,12 +99,15 @@ fn criterion_benchmark(c: &mut Criterion) {
...
+        let inputs = &[
+            ColumnarValue::Array(a3.clone()),
+            ColumnarValue::Array(a4.clone()),
+            ColumnarValue::Scalar(ScalarValue::Int64(Some(42i64))),
+        ];
 
         b.iter(|| {
             for _ in 0..NUM_ITER {
-                create_xxhash64_hashes(&input, &mut dst).unwrap();
+                spark_xxhash64(inputs).unwrap();
```
Comment (kazuyukitanimura) [inline core/benches/hash.rs:110]: is this change related? Looks `spark_xxhash64` covers more code, so performance numbers may change...

**Review by kazuyukitanimura — COMMENTED** [review]
Looks good. One minor question

**Review by kazuyukitanimura — APPROVED** [review]
LGTM

**Review by comphead — APPROVED** [review]
lgtm thanks @advancedxy

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Merged. Thanks @advancedxy @kazuyukitanimura @tshauck @comphead
